## Lists(Списки)

Списки задаются при помощи скобок []

In [2]:
# Создание списка
z = [3, 7, 4, 2]
z

[3, 7, 4, 2]

### Доступ к элементам списка

In [ ]:
#Индексация списка всегда начинается с 0
#Возьмеме первый элемент списка
# 0 1 2 3
z[0]

In [ ]:
#Возьмеме последний элемент списка
z[3]

In [ ]:
#Индексация может идти в обратную сторону, но начинается с -1
#-4 -3 -2 -1

#Возьмем последний элемент списка
z[-1]

In [ ]:
#Функция len позволяет узнать длину списку
len(z)

In [3]:
#поиск по индексу которого нет в списке вернет ошибку
z[100]

IndexError: list index out of range

In [ ]:
#Возьмем последний элемент списка. Код заведомо написан не правильно. Твоя задача его исправить
z[len(z)]

### Slicing Lists

In [ ]:
#Можно выделить часть списка при оператора slice - [a:b:c], где a -индекс с которого начать(включительно), 
#b - индекс которым закончится списко (невключая), c шаг - с которым берутся индексы между a и b

# c нелевого по второй индекс(не включая второй индекс)
z[0:2]

In [ ]:
#Все до 3-го индекса
z[:3]

In [ ]:
#C первого индекса и до конца
z[1:]

In [19]:
#все элементы - таким образом можно копировать списки
z[:]

[3, 7, 4, 2]

In [16]:
#каждый второй элемент
z[1::2]

[7, 2]

In [18]:
#развернуть список
z[::-1]

[2, 4, 7, 3]

### Операции над списками

In [ ]:
#минимум
min(z)

In [ ]:
#максимум
max(z)

In [ ]:
#сумма значений в списке
sum(z)

In [ ]:
#Расчитать сколько раз элемент встретился в списке
random_list = [4, 1, 5, 4, 10, 4]
random_list.count(4)

In [ ]:
#вернуть первый первый индекс элемента
random_list.index(4)

In [ ]:
# you can specify where you start your search. list.index(value, start)
random_list.index(4, 3)

In [ ]:
#Можно задать начало поиска и конец
#list.index(value, [start, stop])
random_list.index(4, 5, 6)


## Сортировка списков

In [78]:
x = [3, 7, 2, 11, 8, 10, 4]
y = ['Steve', 'Rachel', 'Michael', 'Adam', 'Monica', 'Jessica', 'Lester']

In [ ]:
# Сортировка с имзменением исходного списка
# low to high
x.sort()
print(x)

In [ ]:
# high to low
x.sort(reverse = True)
print(x)

In [ ]:
# Сортировка строк
# A-Z
y.sort()
print(y)

In [ ]:
# Z-A
y.sort(reverse = True)
print(y)

In [ ]:
# sсортировка без изменения исходного списка
y = ['Steve', 'Rachel', 'Michael', 'Adam', 'Monica', 'Jessica', 'Lester']

new_list = sorted(y)
new_list

In [ ]:
#Cам y не изменился
y

### Изменение списка

In [77]:
x

257

In [8]:
#Можно просто заменить элемент списка по индексу
x[2] = 10
x

[3, 7, 10, 10, 8, 10, 4]

In [ ]:
#Вставка элемента в конец списка
x.append(12)
x

In [ ]:
#Удалить первый встретившийся элемент из списка
x.remove(10)
x

In [ ]:
#Забрать элемент из списка по индексу
x.pop(3)

In [ ]:
x

In [ ]:
#Соеденить два списка (расширить)
x.extend([3, 6, 9])
x

In [79]:
# оператор  in позволяет проверить есть ли значение в списке
if 3 in x:
    print("3 in list")

3 in list


Задание
1.Cписок состоит из 10 элементов.
Переместить три последних элемента из конца в начало списка без нарушения их исходного порядка.

In [ ]:
x = [1,2,3,4,5,6,7,8,9,10]


In [80]:
# Задание 2
# [1,2,3,4,5,6,7,8,9,10] -> [1,0,3,0,5,0,7,0,8,0] 

In [ ]:
x = [1,2,3,4,5,6,7,8,9,10]

Можно попробовать другие методы работы со списками

Метод|Что делает
---|---
list.append(x)|	Добавляет элемент в конец списка
list.extend(L)|	Расширяет список list, добавляя в конец все элементы списка L
list.insert(i, x)|	Вставляет на i-ый элемент значение x
list.remove(x)|	Удаляет первый элемент в списке, имеющий значение x. ValueError, если такого элемента не существует
list.pop([i])|	Удаляет i-ый элемент и возвращает его. Если индекс не указан, удаляется последний элемент
list.index(x, [start [, end]])|	Возвращает положение первого элемента со значением x (при этом поиск ведется от start до end)
list.count(x)|	Возвращает количество элементов со значением x
list.sort([key=функция])|	Сортирует список на основе функции
list.reverse()|	Разворачивает список
list.copy()|	Поверхностная копия списка
list.clear()|	Очищает список

# Mutable and imutable objects

На этом этапе необходимо ввести понятия изменеяемые(mutable) и не изменяемые объекты(imutable)

Все в Python представлено в виде объектов. Числа, строки, списки, функции, модули - все есть объект.

Объекты могут быть изменяемые, т.е. вы можете изменить их содержимое во время исполнения программы и не изменяемые

In [28]:
#Тип объекта можно узнать при помощи функции type()
#Не изменяемые типы, которые мы уже узнали

print(type(True))
print(type(1))
print(type(3.14))
print(type('hello world'))

<class 'bool'>
<class 'int'>
<class 'float'>
<class 'str'>


In [30]:
#другие immutable типы
print(type((0,0)))
print(type(frozenset("abc")))

<class 'tuple'>
<class 'frozenset'>


In [36]:
#Изменяемые(mutable) типы
print(type([]))
print(type({}))
print(type({"a","b"}))

<class 'list'>
<class 'dict'>
<class 'set'>


In [37]:
#функция id() возвращает идентификатор объекта. В базовой реализации python это номер ячейки памяти в которой хранится значение
x = 10
id(x)

140605988815376

In [45]:
#Посмотрим как ведут себя не изменяемые типы
x = 7891
print(id(x))
y = x
print(id(y))
print(id(x)==id(y))

140605327657872
140605327657872
True


т.е при операции y = x не создается новый объект, а на ячейку в памяти в которой хранится число 7891 ссылается новая переменная.

In [46]:
#А теперь попытаемся изменить значение в переменной x
print(id(x))
x = x+1
print(id(x))
print(x)

140605327657872
140605327658224
7892


теперь x ссылается на новый объект, который хранит значение 7892, и его id совсем другой

In [60]:
#На самом деле дапазон целых чисел -5:256 всегда хранится в памяти при запуске python
x=-5
y=-5
print(id(x)==id(y))

x=256
y=256
print(id(x)==id(y))

True
True


In [61]:
#Значения вне этого диапазона создаются в памяти при каждом новом обращении
x = -6
y = -6
print(id(x)==id(y))

x = 257
y = 257
print(id(x)==id(y))

False
False


Теперь посмотрим на изменяемые объекты.

In [71]:
l = [0,1,2,3,4]
print(l)
id(l)

[0, 1, 2, 3, 4]


140605327610432

In [72]:
l[2] = 10
print(l)
print(id(l))

[0, 1, 10, 3, 4]
140605327610432


Сейчас мы изменили один из элементов списка, но идентификатор списка не поменялся.
Бывают задачи, когда нам необходимо оставить исходный список неизменным. Для этого нам нужно скопировать список

In [73]:
#Для копирования списка можно воспользовать slice
l_2 = l[:]
id(l_2)

140605327680448

In [74]:
l_2[2] = 2
print(l)
print(l_2)

[0, 1, 10, 3, 4]
[0, 1, 2, 3, 4]


In [75]:
#другой способ вызов метода copy модуля copy
from copy import copy
new_l = copy(l)
id(l)

140605327610432

Особое внимание стоит уделить строкам несмотря на то, что к элементам строки можно обращаться по индексу и делать slice, строки - не именные элементы

In [81]:
s = "Hello World"

In [ ]:
s.